File: sample-data-repository/src/main/java/nl/haploid/octowight/sample/data/CaptainResourceFactory.java
package nl.haploid.octowight.sample.data;

import nl.haploid.octowight.registry.data.ResourceRoot;
import nl.haploid.octowight.sample.repository.PersonDmo;
import nl.haploid.octowight.sample.repository.RoleDmo;
import nl.haploid.octowight.sample.repository.RoleDmoRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;

@Component
public class CaptainResourceFactory implements ResourceFactory<CaptainResource> {

	@Autowired
	private RoleDmoRepository roleDmoRepository;

	@Override
	@Transactional(readOnly = true)
	public CaptainResource fromResourceRoot(final ResourceRoot resourceRoot) {
		final RoleDmo roleDmo = roleDmoRepository.findOne(resourceRoot.getAtomId());
		if (roleDmo == null) {
			throw new ResourceNotFoundException();
		}
		roleDmo.setAtomLocus(resourceRoot.getAtomLocus());
		final PersonDmo personDmo = roleDmo.getPerson();
		personDmo.setAtomLocus(resourceRoot.getAtomLocus());
		final CaptainResource captainResource = new CaptainResource();
		captainResource.setId(resourceRoot.getResourceId());
		captainResource.setPersonDmo(personDmo);
		captainResource.setRoleDmo(roleDmo);
		captainResource.setVersion(resourceRoot.getVersion());
		return captainResource;
	}
}

File: sample-data-repository/src/main/scala/nl/haploid/octowight/sample/data/CaptainResourceFactory.scala
package nl.haploid.octowight.sample.data

import nl.haploid.octowight.registry.data.{ResourceFactory, ResourceRoot}
import nl.haploid.octowight.sample.repository.RoleDmoRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional

@Component
class CaptainResourceFactory extends ResourceFactory[CaptainResource] {
  @Autowired private[this] val roleDmoRepository: RoleDmoRepository = null

  @Transactional(readOnly = true)
  def fromResourceRoot(resourceRoot: ResourceRoot): Option[CaptainResource] = {
    Option(roleDmoRepository.findOne(resourceRoot.root.id)) match {
      case Some(roleDmo) =>
        roleDmo.setOrigin(resourceRoot.root.origin)
        val personDmo = roleDmo.getPerson
        personDmo.setOrigin(resourceRoot.root.origin)
        val captainResource = new CaptainResource
        captainResource.setId(resourceRoot.resourceId)
        captainResource.setPersonDmo(personDmo)
        captainResource.setRoleDmo(roleDmo)
        captainResource.setVersion(resourceRoot.version)
        Some(captainResource)
      case None => None
    }
  }
}
